packages/wallet/src/TxBuilder/types.ts

```diff
@@ -0,0 +1,126 @@
+    super(TxOutputFailure.MissingRequiredFields);
+  }
+}
+
+export class OutputValidationMinimumCoinError extends CustomError {
+  public constructor(public txOut: PartialTxOut, public outputValidation: OutputValidation) {
+    super(TxOutputFailure.MinimumCoin);
+  }
+}
+
+export class OutputValidationTokenBundleSizeError extends CustomError {
+  public constructor(public txOut: PartialTxOut, public outputValidation: OutputValidation) {
+    super(TxOutputFailure.TokenBundleSizeExceedsLimit);
+  }
+}
+
+export class IncompatibleWalletError extends CustomError {}
+
+export type TxOutValidationError =
+  | OutputValidationMissingRequiredError
+  | OutputValidationMinimumCoinError
+  | OutputValidationTokenBundleSizeError;
+export type TxBodyValidationError = TxOutValidationError | InputSelectionError;
+
+export type Valid<TValid> = TValid & {
+  isValid: true;
+};
+
+export interface Invalid<TError> {
+  isValid: false;
+  errors: TError[];
+}
+
+export interface ValidTxOutData {
+  readonly txOut: Cardano.TxOut;
+}
+
+export type ValidTxOut = Valid<ValidTxOutData>;
+export type InvalidTxOut = Invalid<TxOutValidationError>;
+export type MaybeValidTxOut = ValidTxOut | InvalidTxOut;
+
+export interface OutputBuilder {
+  partialOutput: PartialTxOut;
+  value(value: Cardano.Value): OutputBuilder;
+  coin(coin: Cardano.Lovelace): OutputBuilder;
+  assets(assets: Cardano.TokenMap): OutputBuilder;
+  /**
+   * @param quantity To remove an asset, set quantity to 0
+   */
+  asset(assetId: Cardano.AssetId, quantity: bigint): OutputBuilder;
+  address(address: Cardano.Address): OutputBuilder;
+  datum(datum: Cardano.util.Hash32ByteBase16): OutputBuilder;
+  build(): Promise<MaybeValidTxOut>;
+}
+
+export interface SignedTx {
+  readonly tx: Cardano.NewTxAlonzo;
+  submit(): Promise<void>;
+}
+
+export interface ValidTxBody {
+  readonly body: Cardano.NewTxBodyAlonzo;
+  readonly auxiliaryData?: Cardano.AuxiliaryData;
+  readonly extraWitness?: Partial<Cardano.Witness>;
+
+  sign(props?: SignTransactionOptions): Promise<SignedTx>;
+}
+
+export type ValidTx = Valid<ValidTxBody>;
+export type InvalidTx = Invalid<TxBodyValidationError>;
+export type MaybeValidTx = ValidTx | InvalidTx;
+
+export interface TxBuilder {
+  partialTxBody: Partial<Cardano.NewTxBodyAlonzo>;
+  auxiliaryData?: Cardano.AuxiliaryData;
+  extraWitness?: Partial<Cardano.Witness>;
+
+  addOutput(txOut: Cardano.TxOut): TxBuilder;
+  /**
+   * @param txOut must be in partialTxBody.outputs (===)
+   */
+  removeOutput(txOut: Cardano.TxOut): TxBuilder;
+  /**
+   * Does *not* addOutput
+   */
+  buildOutput(txOut?: PartialTxOut): OutputBuilder;
+  /**
+   * Add StakeDelegation and (if needed) StakeKeyRegistration certificate
+   *
+   * @throws IncompatibleWalletError when no reward accounts are found
+   */
+  delegate(poolId: Cardano.PoolId): Promise<TxBuilder>;
+  setMetadata(metadata: Cardano.TxMetadata): TxBuilder;
+  build(): Promise<MaybeValidTx>;
+
+  // TODO:
+  // - setMint
```

Comment:
This could also be useful and should be straightforward to implement

```suggestion
  // - setMint
  // - addMetadata(label: bigint, metadatum: Cardano.Metadatum);
  // - removeMetadata(label: bigint);
```

Reply:
Do you mean both `setMint` and `setMetadatum`, or just `setMetadatum`?

Reply:
[9b5e7cb](https://github.com/input-output-hk/cardano-js-sdk/pull/396/commits/9b5e7cbbb5fcb0d2de9dea91ff28c8ce4ae02511)